Add named parameters support

julienfalque opened the pull request:
Implements #594.

theofidry commented:
Nice! Let me know when you think you're done & ready for review :)

julienfalque commented:
Actually I'd like to have your feedback :)

Especially regarding the new `NamedArgumentsResolver` class:
* is its contract ok?
* should it have an interface?

theofidry commented on src/Loader/NativeLoader.php:
```diff
@@ -626,7 +629,7 @@ protected function createPropertyAccessor(): PropertyAccessorInterface
     protected function createCaller(): CallerInterface
     {
-        return new SimpleCaller($this->getCallProcessor(), $this->getValueResolver());
+        return new SimpleCaller($this->getCallProcessor(), $this->getValueResolver(), new NamedArgumentsResolver());
```
I would add a `createNamedArgumentResolver()` method and add `@method NamedArgumentsResolver get NamedArgumentsResolver()`

theofidry commented on src/Generator/Instantiator/Chainable/NoCallerMethodCallInstantiator.php:
```diff
@@ -15,9 +15,20 @@
+     */
+    private $namedArgumentsResolver;
+
+    public function __construct(NamedArgumentsResolver $namedArgumentsResolver = null)
```
```suggestion
    // TODO: make $namedArgumentsResolver non-nullable in 4.0. It is currently nullable only for BC purposes
    public function __construct(NamedArgumentsResolver $namedArgumentsResolver = null)
```

julienfalque replied:
Also added in `SimpleCaller`.

theofidry commented on src/Generator/Instantiator/Chainable/StaticFactoryInstantiator.php:
```diff
@@ -16,10 +16,21 @@
+     */
+    private $namedArgumentsResolver;
+
+    public function __construct(NamedArgumentsResolver $namedArgumentsResolver = null)
```
Same TODO here

theofidry commented on src/Generator/NamedArgumentsResolver.php:
```diff
@@ -0,0 +1,85 @@
+
+class NamedArgumentsResolver
+{
+    public function resolveArguments(array $arguments, string $className, string $methodName)
```
```suggestion
    public function resolveArguments(array $arguments, string $className, string $methodName): array
```

?

theofidry commented on tests/Generator/NamedArgumentsResolverTest.php:
```diff
@@ -0,0 +1,304 @@
+            ],
+        ];
+
+        yield 'with method that does not exist' => [
```
AFAICT it's also missing the case where you are passing the numeric keys as strings e.g.`'0'` which is not uncommon if you need a flag e.g. `0 (unique)`

mlavrinenko commented:
@julienfalque excellent work!

I though about using the fork for this feature, but faced a problem in a case like this:

```php
        yield 'with null as passed argument that have nullable value' => [
            DummyWithMethods::class,
            'methodWithNullables',
            [
                'baz1' => 'hello',
                'baz2' => null,
            ],
            [
                'baz1' => 'hello',
                'baz2' => null,
            ],
        ];

    public function methodWithNullables(?string $baz1, ?string $baz2)
    {
        $this->baz1 = $baz1;
        $this->baz2 = $baz2;
    }
```

It throws:
> Nelmio\Alice\Generator\NamedArgumentsResolverTest::testResolveArguments with data set "with null as passed agrument that have default nullable value" ('Nelmio\Alice\Entity\DummyWithMethods', 'methodWithNullables', array('hello', null), array('hello', null))
RuntimeException: Argument baz2 of Nelmio\Alice\Entity\DummyWithMethods::methodWithNullables() is not passed a value and does not define a default one.

Also I would be glad to use it with inheritance ability, but this doesn't work:

```php
namespace Some;
class Entity {
  public function __construct(string $foo, string $bar): {}
}
```
```yaml
Some\Entity:
  entity_template (template):
    __construct:
      foo: 'value 1'
  entity (extends entity_template):
    __construct:
      bar: 'value 2'
```

julienfalque commented:
@mlavrinenko Good catch! I added the failing test case and fixed it, thanks!

> I would be glad to use it with inheritance ability, but this doesn't work

I'm not familiar with this feature but as far as I understand, it does not support method calls, does it?

mlavrinenko commented:
> I'm not familiar with this feature but as far as I understand, it does not support method calls, does it?

I suppose so. I just wanted to mention that named parameters support will be extra useful with the ability to inherit a template.
Inheritance for __calls could be ambigous and not possible at all, but for __construct it would be fine.

theofidry commented:
@julienfalque is there anything left to do/check or is this good to go?

julienfalque commented:
@theofidry For me the PR is ready for merge unless you think we should work on inheritance as suggested by @mlavrinenko, but I guess that could be a separate PR.

theofidry commented:
No indeed I think this is another feature. Ok I’ll have a look to merge it
before the next release

On Fri 15 May 2020 at 11:46, Julien Falque <notifications@github.com> wrote:

> @theofidry <https://github.com/theofidry> For me the PR is ready for
> merge unless you think we should work on inheritance as suggested by
> @mlavrinenko <https://github.com/mlavrinenko>, but I guess that could be
> a separate PR.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/nelmio/alice/pull/1026#issuecomment-629141427>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABHPVAI7EEWY3WRZ5DTP5SDRRUFPDANCNFSM4L3R2E6Q>
> .
>

julienfalque commented:
I just pushed a small tweak in error messages: argument names now include the `$` sign, this increases readability IMO.

julienfalque commented:
Mmh not sure test failures are related :/

theofidry commented on tests/Loader/LoaderIntegrationTest.php:
```diff
@@ -1978,6 +1978,61 @@ public function provideLegacyFixturesToInstantiate()
             GenerationThrowable::class,
         ];
+
+        yield 'with constructor named parameters' => [
```
I think this is a little error of a merge. The current cases have been appended to `provideLegacyFixturesToInstantiate()` in which the deprecation message is expected (hence the CI failure). They should instead be moved in `provideFixturesToInstantiate()`.

I also notice i did a mistake with a PR earlier and there is still commented cases in `provideFixturesToInstantiate()`. If you can remove them while you at it it would be amazing

julienfalque replied:
Ah thanks for the hint! I overlooked this when rebasing. I removed commented out code too (as a dedicated commit).

theofidry commented:
Hm looks like one test became flaky under 7.4 :(

julienfalque commented:
Thanks @theofidry and @mlavrinenko!

theofidry commented:
Thank you @julienfalque :)
